new routine for finite source size magnification computations

dangilman opened the pull request:
I added a new method in LensModelExtensions that computes magnifications with finite source sizes much faster than the existing magnification_finite routine

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+        return np.array(magnifications)
+
+    @staticmethod
+    def _iterate_adaptive(flux_array, x_image, y_image, grid_x, grid_y, grid_r, r_min, r_max,
```
the naming of this function is a bit misleading. This routine does not do any 'iterative' or 'adaptive' calls

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+    def _iterate_adaptive(flux_array, x_image, y_image, grid_x, grid_y, grid_r, r_min, r_max,
+                          lensModel, kwargs_lens, source_model, kwargs_source):
+        """
+        performs a signle iteration of the ray tracing computations performed in magnification_finite_adaptive
```
here I would specify better what this function does (you don't necessarily need to mention the iterative calling of this function in another definition, as this is part of the other definition)

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+        :param kwargs_lens: keyword arguments for the lens model
+        :param source_fwhm_pasec: the size of the background source in parsecs
+        :param z_source: the source redshift
+        :param astropy: (optional) an instance of astropy; if not specified, a default cosmology will be used
```
to be consistent with the rest of lenstronomy, I would name this variable 'cosmo', in fact it should be an astropy.cosmology instance, and not an astropy instance

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+        """
+
+        if astropy is None:
+            background_cosmology = Background()
```
I would directly make the instance through astropy rather than going through the Background() class to extract the cosmology from there. Makes it more compatible with astropy directly.

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+        The default settings for the grid resolution and ray tracing window size work well for sources with fwhm between
+        0.5 - 100 pc.
+
+        :param x_pos: a list or array of x coordinates
```
in this function, you are using units (translating angle to physical scales with astropy). Can you add the units (arc seconds or pc) in the input requirements for the different parameters?

dangilman replied:
when you say add them to the input requirements, do you mean remove the astropy instance creation from the code and instead explicitly specify a value for parsec_per_arcsec in the function call?

sibirrer replied:
no, I mean that the units of the deflections and angles must be arc seconds, such that the conventions are met. Just add [in arc seconds] or so to the input parameter where this is relevant.

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+            step = step_size * grid_radius_arcsec
+            r_min = 0.
+            r_max = r_min + step
+            flux_array = self._iterate_adaptive(flux_array, xi, yi, grid_x, grid_y, grid_r,
```
these lines of code are kind of duplicated in the while loop. Is it possible that also the first calculation is already part of the while loop to make it more compact?

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+                                      astropy=None, grid_resolution=None, grid_radius_arcsec=None, tol=0.001,
+                                      grid_res_scale=0.0004, grid_size_scale=0.004, step_size=0.05):
+        """
+        This method computes image magnifications with a finite-size background source assuming a Gaussian
```
can you also describe a bit more the iterative procedure in this scheme?

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+                                      grid_res_scale=0.0004, grid_size_scale=0.004, step_size=0.05):
+        """
+        This method computes image magnifications with a finite-size background source assuming a Gaussian
+        source light profile. This method can be much faster that magnification_finite for lens models with many
```
Here you can also specify that this effectively a speed-up in general when the ray-tracing is the bottle neck and most source positions are zeros.

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +25,143 @@ def hessian()
+        source_model = LightModel(['GAUSSIAN'])
+
+        npix = int(2 * grid_radius_arcsec / grid_resolution)
+        _grid_x = np.linspace(-grid_radius_arcsec, grid_radius_arcsec, npix)
```
could there be a further speed-up if the grid is not rectangular and already tight to the magnification tensor of the lensing feature? Or do you want to deliberately not have it be that way to have better non-local distortion control? Might be also worth mentioning in the description (sorry to make you write a lot of text)

dangilman replied:
I initially intended to do that but it's a bit more complicated and I didn't feel like coding it up at the time. A circular aperture is also 'safe' in the sense that it should capture highly-curved images that might extended beyond an ellipse estimated from the magnification tensor

sibirrer replied:
ok, that's fine. But it's useful still to clarify that in the top-level documentation such that if someone else picks it up, there is space for extensions

dangilman replied:
ok, if I were to include it now, would it basically amount to rotating the ellipse such that the major axis is parallel to the eigenvector with the largest eigenvalue?

sibirrer replied:
yes (probably), but if you are not using it, it would require you to do also quite some tests to make sure it's working. If you make a comment/TODO/idea that'll do.

dangilman replied:
I think I got it working, I make some plots of the full ray tracing through the rectangular window the ray tracing through an ellipse with major axis parallel to largest eigenvector 

![image](https://user-images.githubusercontent.com/15677070/104537779-ff612280-55f0-11eb-8a97-1e75a083489c.png)

dangilman replied:
![image](https://user-images.githubusercontent.com/15677070/104537793-07b95d80-55f1-11eb-80fc-d9c8a5f4e877.png)

![image](https://user-images.githubusercontent.com/15677070/104537807-0c7e1180-55f1-11eb-817f-0331668557fe.png)

![image](https://user-images.githubusercontent.com/15677070/104537845-1acc2d80-55f1-11eb-8022-5bfec14067a5.png)

sibirrer commented on test/test_LensModel/test_lens_model_extensions.py:
```diff
@@ -75,12 +76,63 @@ def test_elliptical_ray_trace(self):
+        x_image, y_image = solver.findBrightImage(source_x, source_y, kwargs_lens)
+
+        source_fwhm_parsec = 60.
+        background = Background()
```
can you make the test with a specific astropy.cosmology instance? In this form, if the default cosmology is changing (not in control of lenstronomy) this tests could lead to different results.

sibirrer commented on test/test_LensModel/test_lens_model_extensions.py:
```diff
@@ -75,12 +76,63 @@ def test_elliptical_ray_trace(self):
+
+        mag_square_grid *= max(mag_square_grid) ** -1
+        mag_adaptive_grid *= max(mag_adaptive_grid) ** -1
+        npt.assert_almost_equal(mag_square_grid, mag_adaptive_grid, 3)
```
here you test against magnification_finite(), which is great! Can you also test against the expected magnification, given the knowledge of the (local) magnification and the intrinsic source brightness? (I know it's not perfect due to the extended size of the source, but perhaps what you can test is with a constant hessian lens model (there you know the answer of the constant magnification over an extended range)

dangilman replied:
I'm not really sure what you mean

sibirrer replied:
a test for the absolute magnification/brightness relative to the (analytically known) expected brightness derived from intrinsic brightness x magnification

dangilman replied:
so basically a test w.r.t. lensmodel.magnification?

sibirrer replied:
yes, correct!

sibirrer commented on test/test_LensModel/test_lens_model_extensions.py:
```diff
@@ -75,12 +76,63 @@ def test_elliptical_ray_trace(self):
+        mag_adaptive_grid = extension.magnification_finite_adaptive(x_image, y_image, source_x, source_y, kwargs_lens, source_fwhm_parsec,
+                                                                    z_source)
+
+        mag_square_grid *= max(mag_square_grid) ** -1
```
testing against magnification_finite does not require you to re-scale with the magnification. Perhaps you want to test this in relative precision rather than in absolute one (modulo magnification).

sibirrer commented on test/test_LensModel/test_lens_model_extensions.py:
```diff
@@ -75,12 +76,63 @@ def test_elliptical_ray_trace(self):
+        r_max = source_sigma*0.9
+        flux_array = extension._iterate_adaptive(flux_array, x_image, y_image, grid_x, grid_y, grid_r, r_min, r_max,
+                          lensmodel, kwargs_lens, source_model, kwargs_source)
+        npt.assert_equal(True, flux_array[0] > 0.)
```
is there a more specific expected outcome of this?

sibirrer reviewed (changes requested):
Thanks for making this PR! This is a nice addition and well documented. I added some requests for further documentation and minor changes of the design and tests. Overall, this function matches the type of functionality in the LensModelExtension routines.

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -1,6 +1,9 @@
 import numpy as np
 import lenstronomy.Util.util as util
 from skimage.measure import find_contours
+from lenstronomy.Cosmo.background import Background
```
do you need that still?

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -1,6 +1,9 @@
 from skimage.measure import find_contours
+from lenstronomy.Cosmo.background import Background
+from lenstronomy.LightModel.light_model import LightModel
+from astropy.cosmology import default_cosmology
```
perhaps to make the imports a bit more lightweight on top, can you import this directly just before using the cosmology? It's a class instance so it might be used multiple times (at least in theory)

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +26,149 @@ def hessian()
+        be estimated from the source size
+        :param grid_radius_arcsec: (optional) the size of the ray tracing region; if not specified, an appropriate value
+        will be estimated from the source size
+        :param axis_ratio: the axis ratio of the ellipse used for ray tracing
```
do you also want to have the option such that the axis ratio is the tangential/radial eigenvalue? This might further optimize (if local value is reflective to the non-local behavior)

sibirrer commented on lenstronomy/LensModel/lens_model_extensions.py:
```diff
@@ -23,6 +26,149 @@ def hessian()
+                v = np.array([v21, v22])
+            rotation_angle = np.arctan(v[1] / v[0]) - np.pi / 2
+            grid_x, grid_y = util.rotate(grid_x_0, grid_y_0, rotation_angle)
+            grid_r = np.hypot(grid_x, grid_y / axis_ratio).ravel()
```
the axis ratio might be also used from the eigenvalues w1 and w2 (same comment as above)

sibirrer commented on test/test_LensModel/test_lens_model_extensions.py:
```diff
@@ -6,15 +6,18 @@
 from lenstronomy.LensModel.lens_model_extensions import LensModelExtensions
 from lenstronomy.LensModel.Solver.lens_equation_solver import LensEquationSolver
 from lenstronomy.LensModel.lens_model import LensModel
+from lenstronomy.Cosmo.background import Background
```
do you need that import still?

sibirrer commented on test/test_LensModel/test_lens_model_extensions.py:
```diff
@@ -75,12 +77,76 @@ def test_elliptical_ray_trace(self):
+
+        mag_point_source = abs(lensmodel.magnification(x_image, y_image, kwargs_lens))
+
+        npt.assert_almost_equal(mag_square_grid/mag_adaptive_grid, np.ones_like(mag_square_grid), 2)
```
this test is to the percent level. Does the test fail at the per mill level? Sub-percent precision might be necessary for very precise flux measurements and it's better it fulfills this requirement.

sibirrer reviewed (changes requested):
two more minor requests and then I think we are good to go. Thanks a lot for putting your time into this!

dangilman commented:
Hi,

I think I addressed all your comments in the version I just pushed. 

About the precision, the magnifications are all off by a constant factor, which means that the smaller ray tracing area in the new routine misses a small fraction of the total integrated flux. This can be fixed by making the default grid size larger, but that in turn slows down the algorithm a bit.  

I made a compromise by increasing the default grid size such that the magnifications are accurate to within half a percent. Since the discrepancy is a constant factor, the flux ratios are more precise, accurate to a quarter percent with the default setting. 

- Daniel 

> On Jan 14, 2021, at 12:44 AM, Simon Birrer <notifications@github.com> wrote:
> 
> 
> @sibirrer requested changes on this pull request.
> 
> two more minor requests and then I think we are good to go. Thanks a lot for putting your time into this!
> 
> In lenstronomy/LensModel/lens_model_extensions.py <https://github.com/sibirrer/lenstronomy/pull/231#discussion_r557038333>:
> 
> > @@ -1,6 +1,9 @@
>  import numpy as np
>  import lenstronomy.Util.util as util
>  from skimage.measure import find_contours
> +from lenstronomy.Cosmo.background import Background
> do you need that still?
> 
> In lenstronomy/LensModel/lens_model_extensions.py <https://github.com/sibirrer/lenstronomy/pull/231#discussion_r557038642>:
> 
> > @@ -1,6 +1,9 @@
>  import numpy as np
>  import lenstronomy.Util.util as util
>  from skimage.measure import find_contours
> +from lenstronomy.Cosmo.background import Background
> +from lenstronomy.LightModel.light_model import LightModel
> +from astropy.cosmology import default_cosmology
> perhaps to make the imports a bit more lightweight on top, can you import this directly just before using the cosmology? It's a class instance so it might be used multiple times (at least in theory)
> 
> In lenstronomy/LensModel/lens_model_extensions.py <https://github.com/sibirrer/lenstronomy/pull/231#discussion_r557039242>:
> 
> > +        you can just set axis_ratio = 1.
> +
> +        The default settings for the grid resolution and ray tracing window size work well for sources with fwhm between
> +        0.5 - 100 pc.
> +
> +        :param x_image: a list or array of x coordinates [units arcsec]
> +        :param y_image: a list or array of y coordinates [units arcsec]
> +        :param kwargs_lens: keyword arguments for the lens model
> +        :param source_fwhm_parsec: the size of the background source [units parsec]
> +        :param z_source: the source redshift
> +        :param cosmo: (optional) an instance of astropy.cosmology; if not specified, a default cosmology will be used
> +        :param grid_resolution: the grid resolution in units arcsec/pixel; if not specified, an appropriate value will
> +        be estimated from the source size
> +        :param grid_radius_arcsec: (optional) the size of the ray tracing region; if not specified, an appropriate value
> +        will be estimated from the source size
> +        :param axis_ratio: the axis ratio of the ellipse used for ray tracing
> do you also want to have the option such that the axis ratio is the tangential/radial eigenvalue? This might further optimize (if local value is reflective to the non-local behavior)
> 
> In lenstronomy/LensModel/lens_model_extensions.py <https://github.com/sibirrer/lenstronomy/pull/231#discussion_r557039637>:
> 
> > +
> +        magnifications = []
> +        minimum_magnification = 1e-4
> +        grid_x_0, grid_y_0 = np.meshgrid(_grid_x, _grid_y)
> +        grid_x_0, grid_y_0 = grid_x_0.ravel(), grid_y_0.ravel()
> +
> +        for i, (xi, yi) in enumerate(zip(x_image, y_image)):
> +
> +            w1, w2, v11, v12, v21, v22 = self.hessian_eigenvectors(xi, yi, kwargs_lens)
> +            if abs(w1) > abs(w2):
> +                v = np.array([v11, v12])
> +            else:
> +                v = np.array([v21, v22])
> +            rotation_angle = np.arctan(v[1] / v[0]) - np.pi / 2
> +            grid_x, grid_y = util.rotate(grid_x_0, grid_y_0, rotation_angle)
> +            grid_r = np.hypot(grid_x, grid_y / axis_ratio).ravel()
> the axis ratio might be also used from the eigenvalues w1 and w2 (same comment as above)
> 
> In test/test_LensModel/test_lens_model_extensions.py <https://github.com/sibirrer/lenstronomy/pull/231#discussion_r557039840>:
> 
> > @@ -6,15 +6,18 @@
>  from lenstronomy.LensModel.lens_model_extensions import LensModelExtensions
>  from lenstronomy.LensModel.Solver.lens_equation_solver import LensEquationSolver
>  from lenstronomy.LensModel.lens_model import LensModel
> +from lenstronomy.Cosmo.background import Background
> do you need that import still?
> 
> In test/test_LensModel/test_lens_model_extensions.py <https://github.com/sibirrer/lenstronomy/pull/231#discussion_r557040162>:
> 
> > +
> +        mag_adaptive_grid = extension.magnification_finite_adaptive(x_image, y_image, source_x, source_y, kwargs_lens, source_fwhm_parsec,
> +                                                                    z_source, cosmo=self.cosmo, tol=0.0001)
> +
> +        # tests the default cosmology
> +        _ = extension.magnification_finite_adaptive(x_image, y_image, source_x, source_y, kwargs_lens,
> +                                                                    source_fwhm_parsec,
> +                                                                    z_source, cosmo=None, tol=0.0001)
> +        # tests the r_max > sqrt(2) * grid_radius stop criterion
> +        _ = extension.magnification_finite_adaptive(x_image, y_image, source_x, source_y, kwargs_lens,
> +                                                    source_fwhm_parsec,
> +                                                    z_source, cosmo=None, tol=0.0001, step_size=1000)
> +
> +        mag_point_source = abs(lensmodel.magnification(x_image, y_image, kwargs_lens))
> +
> +        npt.assert_almost_equal(mag_square_grid/mag_adaptive_grid, np.ones_like(mag_square_grid), 2)
> this test is to the percent level. Does the test fail at the per mill level? Sub-percent precision might be necessary for very precise flux measurements and it's better it fulfills this requirement.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/sibirrer/lenstronomy/pull/231#pullrequestreview-567845548>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADXTNDUUHE3OU2FM6B7UCCDSZZZBXANCNFSM4WBRXCHA>.
>

sibirrer reviewed (approved):
Thank you Daniel! This looks all good to me! One note on the cosmology dependence that came to my mind: The LensModel instance might also be cosmology dependent and in the way this routine is coded, the user might accidentally use two different cosmologies. So in the long run either asking the user to pass the angular source size (meaning putting more responsibility directly in the hands of the user) or having a more generalized 'default cosmology' within the lenstronomy environment might be useful. Just some thoughts.
